### PR TITLE
Fix error when float variable value is greater 10^15-1

### LINKED ISSUE #47: Wrong variable "var" for the Scalar "Float" if value is greater than 10^15

### PR DESCRIPTION
**Pull request by no1seman**
According to http://www.lua.org/pil/2.3.html before this patch if mantissa is greater than 100,000,000,000,000 tonumber converts value not to number but to cdata(int64_t or uint64_t) and previous algorithm checked value type only for "number". If value is greater that 10^15-1 there was an error because type(value) ~= 'number'. This patch fixes such misbehavour.

closes: https://github.com/tarantool/graphql/issues/47

### CODE REVIEW

File: graphql/types.lua [hunk 1]
```diff
@@ -318,7 +318,19 @@ types.long = types.scalar({
 })
 
 local function isFloat(value)
-  return type(value) == 'number'
+  if type(value) == 'number' then
+    return true
+  end
+
+  if type(value) == 'cdata' then
```
Comment (olegrok) [inline graphql/types.lua]: This check looks redundant. `ffi.istype('int64_t', value)` will work with any values.

E.g.
```
tarantool> ffi.istype('int64_t', 'abc')
---
- false
...
```

File: test/integration/graphql_test.lua [hunk 2]
```diff
@@ -89,14 +89,24 @@ end
...
+        }
     ]]
-    local variables = {arg = 'B', arg2 = '22'}
+    local variables = json.decode([[
+        {
+            "arg": "B",
+            "arg2": "22",
+            "arg3": 2147483647,
+            "arg4": 9223372036854775807,
+            "arg5": 9007199254740991
```
Comment (olegrok) [inline test/integration/graphql_test.lua]: Please extend test for float values. Bot uint64_t and int64_t should be tested. Also "negative" cases should be tested as well (when value >= 2^53)

File: graphql/types.lua [hunk 3]
```diff
@@ -318,7 +318,19 @@ types.long = types.scalar({
...
+    if ffi.istype('int64_t', value) then
+      return value >= -2^53 and value < 2^53
+    elseif ffi.istype('uint64_t', value) then
+      return value < 2^53
```
Comment (olegrok) [inline graphql/types.lua]: Also I guess we need to be sure that comparison between LL/ULL and Lua numbers produces correct result (https://github.com/tarantool/tarantool/issues/4089). Probably tonumber should be additionally applied

File: test/integration/graphql_test.lua [hunk 4]
```diff
@@ -2051,3 +2051,65 @@ g.test_propagate_defaults_to_callback = function()
     t.assert_equals(errors, nil)
     t.assert_items_equals(json.decode(data.prefix.test_mutation), result)
 end
+
+function g.test_float_variables()
+    local query = [[
+        query ($arg: Float) { test(arg: $arg) }
```
Comment (olegrok) [inline test/integration/graphql_test.lua:2057]: It's relevant if float value passed via variables. Let's test also argument value.

**Review by olegrok — APPROVED** [review]
Thanks. LGTM.

**Comment by Totktonada** [comment]
2^53 or 10^15?

Sure, values over the [-2^53; 2^53] range has precision over 1.0 (not all integer numbers are representable out of this range), but the graphql standard due not obligate or instruct us to impose particular precision. Moreover: it links the IEEE 753 standard, which allows such double-precision floats.

Regarding `inf` and `NaN` -- okay. Regarding large numbers -- unclear for me.

**Comment by no1seman** [comment]
@Totktonada http://www.lua.org/pil/2.3.html <- look

**Comment by Totktonada** [comment]
I already summarized it above.

**Comment by no1seman** [comment]
What do you suggest?

**Comment by Totktonada** [comment]
I don't understand why, say, 2^54 is not considered as a double-precision floating point number. Can you clarify it?

**Comment by no1seman** [comment]
@Totktonada Here (http://spec.graphql.org/October2021/#sec-Float) spec points to https://en.wikipedia.org/wiki/IEEE_754 where Double precision floating point value may have up to 53 Significand bits aka mantissa. Fix me if I'm not right.

**Comment by Totktonada** [comment]
> Double precision floating point value may have up to 53 Significand bits aka mantissa.

Playing with exponent values allows to represent some numbers on the order of 10^308. 53 binary digits in the mantissa does not mean that the range is [-2^53; 2^53-1]. It is just the range, where we have enough precision to represent all integer values (without holes).

You can look how double numbers are represented in the memory and make sure that numbers above 2^53 are actually representable:

```shell
$ cc --std=c11 -Wall -Wextra -x c <(echo -e '#include <stdio.h>\nstatic int idx(int n) { return __FLOAT_WORD_ORDER__ == __ORDER_LITTLE_ENDIAN__ ? 7 - n : n; }\nint main() { double x = 18014398509481984.0; printf("Decimal representation: %lf\\n", x); unsigned char *cs = (unsigned char *)&x; printf("In memory: %02hhX %02hhX %02hhX %02hhX %02hhX %02hhX %02hhX %02hhX\\n", cs[idx(0)], cs[idx(1)], cs[idx(2)], cs[idx(3)], cs[idx(4)], cs[idx(5)], cs[idx(6)], cs[idx(7)]); return 0; }') && ./a.out; rm a.out
Decimal representation: 18014398509481984.000000
In memory: 43 50 00 00 00 00 00 00
```

**Comment by Totktonada** [comment]
@no1seman Thank you for bringing attention on the problems and for the proposal of the fix.

I think that we should accept cdata numbers as floating point values. I proposed this approach in PR #49. Please, take a look if time permits.

**Comment by no1seman** [comment]
@Totktonada I can't see any principal difference in your solution except of remove of cheking of 2^53 for cdata. All other things is a matter of personal taste. Let's close this PR and merge yours.

**Comment by no1seman** [comment]
Closed in favour of https://github.com/tarantool/graphql/pull/49

**Pull request closed without merging**